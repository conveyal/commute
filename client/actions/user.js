import {createAction} from 'redux-actions'

// login / logout
export const login = createAction('log in')
export const logout = createAction('log out')
