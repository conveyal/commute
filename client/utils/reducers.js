export function arrayToObj (a) {
  const obj = {}
  for (let i = 0; i < a.length; i++) obj[a[i].id] = a[i]
  return obj
}

/**
 * Round a number to a fixed amount of decimal places
 *
 * @param  {Number} n      The number to round
 * @param  {Number} places The number of places to round to (default=2)
 * @return {Number}        The rounded number
 */
export function fixedRound (n, places) {
  places = places || 2
  const multiplier = Math.pow(10, places)
  return Math.round(n * multiplier) / multiplier
}
