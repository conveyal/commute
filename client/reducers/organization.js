import update from 'react-addons-update'

import {settings} from '../utils/env'
import {arrayToObj, fixedRound} from '../utils/reducers'

const METRICS = Object.keys(settings.metrics)

export const reducers = {

  /**
   * Organization Stuff
   */

  'add organization' (state, action) {
    return addEntities(state, 'organizations', [action.payload])
  },
  'delete organization' (state, action) {
    return deleteEntity(state, 'organizations', action.payload.id)
  },
  'set organization' (state, action) {
    return setOrganization(state, action.payload)
  },
  'set organizations' (state, action) {
    const organizations = [...action.payload]
    return {
      organizations,
      organizationsById: arrayToObj(action.payload)
    }
  },

  /**
   * Analysis Stuff
   */

  'add analysis' (state, action) {
    const {organizationId, analysis} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = addEntities(affectedOrganization, 'analyses', [analysis])
    return setOrganization(state, updatedOrganization)
  },
  'delete analysis' (state, action) {
    const {organizationId, analysisId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = deleteEntity(affectedOrganization, 'analyses', analysisId)
    return setOrganization(state, updatedOrganization)
  },
  'receive mock calculated trips' (state, action) {
    const {analysisId, organizationId, trips} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const affectedAnalysis = affectedOrganization.analysesById[analysisId]
    let updatedAnalysis = update(affectedAnalysis, {
      trips: { $set: trips }
    })

    // calculate array of values for each metric
    const vals = {}
    const modes = Object.keys(trips[0]).filter((mode) => (['commuterId']).indexOf(mode) === -1)
    modes.forEach((mode) => {
      vals[mode] = {
        cost: [],
        distance: [],
        time: []
      }
    })
    trips.forEach((trip) =>
      modes.forEach((mode) =>
        METRICS.forEach((metric) => {
          vals[mode][metric].push(trip[mode][metric])
        })
      )
    )

    // sort arrays
    modes.forEach((mode) =>
      METRICS.forEach((metric) => {
        vals[mode][metric].sort()
      })
    )

    updatedAnalysis = update(updatedAnalysis, {
      tripVals: { $set: vals }
    })

    // calculate analysis summary metrics
    // Potential Savings is an overall average of the difference in cost between
    // the most likely trip and the driving trip
    const len = trips.length
    let totalCostDifference = 0
    let totalTravelTime = 0
    let totalDistance = 0
    trips.forEach((trip) => {
      totalCostDifference += trip.car.cost - trip.mostLikely.cost
      totalTravelTime += trip.mostLikely.time
      totalDistance += trip.mostLikely.distance
    })

    updatedAnalysis = update(updatedAnalysis, {
      summary: {
        $set: {
          avgTravelTime: fixedRound(totalTravelTime / len),
          avgDistance: fixedRound(totalDistance / len),
          savingsPerTrip: fixedRound(totalCostDifference / len),
          savingsPerTripYear: fixedRound(totalCostDifference / len * 260),
          savingsTotalPerDay: fixedRound(totalCostDifference),
          savingsTotalPerYear: fixedRound(totalCostDifference * 260)
        }
      }
    })

    const updatedOrganization = setEntity(affectedOrganization, 'analyses', updatedAnalysis)
    return setOrganization(state, updatedOrganization)
  },

  /**
   * Commuter Stuff
   */

  'add commuter' (state, action) {
    const {commuter, groupId, organizationId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const affectedGroup = affectedOrganization.groupsById[groupId]
    const updatedGroup = addEntities(affectedGroup, 'commuters', [commuter])
    const updatedOrganization = setGroup(affectedOrganization, updatedGroup)
    return setOrganization(state, updatedOrganization)
  },
  'delete commuter' (state, action) {
    const {commuterId, groupId, organizationId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const affectedGroup = affectedOrganization.groupsById[groupId]
    const updatedGroup = deleteEntity(affectedGroup, 'commuters', commuterId)
    const updatedOrganization = setGroup(affectedOrganization, updatedGroup)
    return setOrganization(state, updatedOrganization)
  },
  'update commuter' (state, action) {
    const {commuter, groupId, organizationId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const affectedGroup = affectedOrganization.groupsById[groupId]
    const updatedGroup = setEntity(affectedGroup, 'commuters', commuter)
    const updatedOrganization = setGroup(affectedOrganization, updatedGroup)
    return setOrganization(state, updatedOrganization)
  },

  /**
   * Group Stuff
   */

  'add group' (state, action) {
    const {organizationId, group} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    group.commutersById = arrayToObj(group.commuters)
    const updatedOrganization = addEntities(affectedOrganization, 'groups', [group])
    return setOrganization(state, updatedOrganization)
  },
  'append commuters' (state, action) {
    const {commuters, groupId, organizationId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const affectedGroup = affectedOrganization.groupsById[groupId]
    const updatedGroup = addEntities(affectedGroup, 'commuters', commuters)
    const updatedOrganization = setGroup(affectedOrganization, updatedGroup)
    return setOrganization(state, updatedOrganization)
  },
  'delete group' (state, action) {
    const {organizationId, groupId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = deleteEntity(affectedOrganization, 'groups', groupId)
    return setOrganization(state, updatedOrganization)
  },

  /**
   * Site Stuff
   */

  'add site' (state, action) {
    const {organizationId, site} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = addEntities(affectedOrganization, 'sites', [site])
    return setOrganization(state, updatedOrganization)
  },
  'delete site' (state, action) {
    const {organizationId, siteId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = deleteEntity(affectedOrganization, 'sites', siteId)
    return setOrganization(state, updatedOrganization)
  },
  'update site' (state, action) {
    const {organizationId, site} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = setEntity(affectedOrganization, 'sites', site)
    return setOrganization(state, updatedOrganization)
  }
}

// Utilities

/**
 * Helper to delete an item from a entityById object
 * @param  {String} id  The ID of the entity to delete
 * @return {Function}   A function that deletes the entity from
 *  the object and then returns that object
 */
const deleteFromMap = (id) => (map) => {
  delete map[id]
  return map
}

export const initialState = {
  organizations: [],
  organizationsById: {}
}

/**
 * Add entities to a collection
 *
 * @param {Object} parent           the parent object containing the collections
 * @param {String} entitiesName     base name for entity collection keys
 * @param {Array} entities          the entities to add
 * @return                          The updated parent
 */
const addEntities = (parent, entitiesName, entities) => {
  const entitiesByIdName = `${entitiesName}ById`
  let updatedParent = update(parent, {
    [entitiesName]: { $push: entities }
  })
  return update(updatedParent, {
    [entitiesByIdName]: { $set: arrayToObj(updatedParent[entitiesName]) }
  })
}

/**
 * Delete an entity within a collection
 *
 * @param {Object} parent           the parent object containing the collections
 * @param {String} entitiesName     base name for entity collection keys
 * @param {Object} entityId         ID of the entity to delete
 * @return                          The updated parent
 */
const deleteEntity = (parent, entitiesName, entityId) => {
  const entitiesByIdName = `${entitiesName}ById`
  let updatedParent = update(parent, {
    [entitiesByIdName]: { $apply: deleteFromMap(entityId) }
  })
  return update(updatedParent, {
    [entitiesName]: { $set: Object.values(updatedParent[entitiesByIdName]) }
  })
}

/**
 * Helper to update an entity within a collection
 *
 * @param {Object} parent           the parent object containing the collections
 * @param {String} entitiesName     base name for entity collection keys* @param {Object} updatedEntity    entity to update
 * @return                          The updated parent
 */
const setEntity = (parent, entitiesName, updatedEntity) => {
  const entitiesByIdName = `${entitiesName}ById`
  let updatedParent = update(parent, {
    [entitiesByIdName]: {
      [updatedEntity.id]: {
        $set: updatedEntity
      }
    }
  })
  return update(updatedParent, {
    [entitiesName]: {
      $set: Object.values(updatedParent[entitiesByIdName])
    }
  })
}

const setGroup = (organization, group) => {
  return setEntity(organization, 'groups', group)
}

const setOrganization = (state, organization) => {
  return setEntity(state, 'organizations', organization)
}
