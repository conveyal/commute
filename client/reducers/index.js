import {handleActions} from 'redux-actions'

import * as organization from './organization'
import * as user from './user'

export default {
  organization: handleActions(organization.reducers, organization.initialState),
  user: handleActions(user.reducers, user.initialState)
}
