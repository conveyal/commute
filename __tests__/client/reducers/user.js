/* globals describe, expect, it */

import {handleActions} from 'redux-actions'

import * as user from '../../../client/reducers/user'

describe('client > reducers > user', () => {
  const reducer = handleActions(user.reducers, user.initialState)

  // Default State Test
  it('should handle default state', () => {
    expect(reducer(undefined, { type: 'blah', payload: {} })).toMatchSnapshot()
  })

  // Specific Handler Tests
  it('should handle log out', () => {
    const action = { type: 'log out', payload: {} }
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set auth0 user', () => {
    const action = { type: 'set auth0 user', payload: { id: 1 } }
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set id token', () => {
    const action = { type: 'set id token', payload: 'abcd' }
    expect(reducer(undefined, action)).toMatchSnapshot()
  })
})
