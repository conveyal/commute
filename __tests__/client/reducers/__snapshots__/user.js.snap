// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client > reducers > user should handle default state 1`] = `Object {}`;

exports[`client > reducers > user should handle log out 1`] = `Object {}`;

exports[`client > reducers > user should handle set auth0 user 1`] = `
Object {
  "id": 1,
}
`;

exports[`client > reducers > user should handle set id token 1`] = `
Object {
  "idToken": "abcd",
}
`;
